File: questPresentation/src/main/java/ru/nekit/android/qls/window/common/QuestWindowMediator.java
package ru.nekit.android.qls.window.common;


import android.content.Intent;
import android.support.annotation.NonNull;
import android.support.annotation.StyleRes;
import android.view.View;

import ru.nekit.android.qls.EventBus;
import ru.nekit.android.qls.quest.QuestContext;
import ru.nekit.android.qls.window.Window;
import ru.nekit.android.qls.window.WindowContentViewHolder;

public abstract class QuestWindowMediator implements EventBus.IEventHandler {

    @NonNull
    protected final QuestContext mQuestContext;
    @NonNull
    private WindowContentViewHolder mWindowContentViewHolder;
    private Window mWindow;
    private View.OnAttachStateChangeListener onAttachStateChangeListener =
            new View.OnAttachStateChangeListener() {
                @Override
                public void onViewAttachedToWindow(View view) {
                    onWindowContentAttach(view);
                }

                @Override
                public void onViewDetachedFromWindow(View view) {

                }
            };

    public QuestWindowMediator(@NonNull QuestContext questContext) {
        mQuestContext = questContext;
    }

    protected abstract WindowContentViewHolder createWindowContent();

    @StyleRes
    protected abstract int getWindowStyleId();

    public final void openWindow() {
        if (mWindow == null) {
            mWindowContentViewHolder = createWindowContent();
            mWindow = new QuestWindow(mQuestContext,
                    "QuestWindowMediator",
                    mWindowContentViewHolder,
                    getWindowStyleId());
            mQuestContext.getEventBus().handleEvents(this, QuestWindow.EVENT_WINDOW_OPEN);
        }
        mWindowContentViewHolder.view.addOnAttachStateChangeListener(onAttachStateChangeListener);
        mWindow.open();
    }

    protected void onWindowContentAttach(View view) {
    }

    protected void destroy() {
    }

    @Override
    public void onEvent(@NonNull Intent intent) {
        mWindowContentViewHolder.view.removeOnAttachStateChangeListener(onAttachStateChangeListener);
    }

    protected void closeWindow(String position) {
        mWindow.close(position);
    }

    @NonNull
    @Override
    public String getEventBusName() {
        return null;
    }
}

File: questPresentation/src/main/java/ru/nekit/android/qls/window/common/QuestWindowMediator.kt
package ru.nekit.android.qls.window.common

import android.support.annotation.CallSuper
import android.support.annotation.StyleRes
import io.reactivex.Single
import io.reactivex.disposables.CompositeDisposable
import ru.nekit.android.qls.quest.QuestContext
import ru.nekit.android.qls.quest.providers.IQuestContextSupport
import ru.nekit.android.window.Window
import ru.nekit.android.window.WindowContentViewHolder

abstract class QuestWindowMediator(override var questContext: QuestContext) : IQuestContextSupport {

    private lateinit var window: Window
    private lateinit var windowContentViewHolder: WindowContentViewHolder

    override var disposableMap: MutableMap<String, CompositeDisposable> = HashMap()

    @get:StyleRes
    protected abstract val windowStyleId: Int

    protected abstract fun createWindowContent(): Single<WindowContentViewHolder>

    fun openWindow(body: () -> Unit) {
        autoDispose {
            createWindowContent().subscribe { it ->
                windowContentViewHolder = it
                QuestWindow(questContext,
                        getName(),
                        windowContentViewHolder,
                        windowStyleId).also {
                    window = it
                    listenForWindowEvent {
                        if (it == QuestWindowEvent.CLOSED)
                            destroy()
                    }
                }.open()
                body()
            }
        }
    }

    abstract fun getName(): String

    @CallSuper
    protected open fun destroy() {
        dispose()
    }

    protected fun closeWindow(position: String) {
        window.close(position)
    }

}